File: src/harness/adapters/qwen.py
"""qwen adapter — invokes the `qwen` CLI in print mode.

`qwen -p PROMPT --output-format json` emits a JSON array on stdout where the
last item with `type='result'` carries usage: `{"input_tokens": N, "output_tokens": M}`.
Alibaba Cloud does not embed pricing in the response.

Interactive sessions are recorded by qwen-code's ChatRecordingService as
`<runtime>/projects/<sanitizeCwd(projectRoot)>/chats/<sessionId>.jsonl`, where
`<runtime>` is `QWEN_RUNTIME_DIR`, else `QWEN_HOME`, else `~/.qwen`, and
`sanitizeCwd` replaces every non-alphanumeric character with `-`. Sanitised
paths can collide, so every record carries the literal `cwd` and discovery
rejects sessions recorded for another project.
"""
from __future__ import annotations

import json
import os
import re
from pathlib import Path

from harness._subproc import SubprocOutcome
from harness.base import (
    Adapter,
    AgentStatus,
    BuildCommand,
    InstallMeta,
    ParsedOutput,
    ReadyState,
    RunSpec,
    SessionTelemetry,
)
from harness.pricing import derive_cost
from harness.util import last_non_empty_join

_AUTH_RE = re.compile(r"Qwen OAuth|API Key", re.IGNORECASE)
_AUTH_ACTION_RE = re.compile(r"Discontinued|switch", re.IGNORECASE)
_PROMPT_RE = re.compile(r"Type your message|>\s*$|❯\s*$", re.IGNORECASE | re.MULTILINE)
_RATE_LIMIT_RE = re.compile(r"rate.?limit|quota", re.IGNORECASE)
_SPINNER_RE = re.compile(r"[⠋⠙⠹⠸⠼⠴⠦⠧⠇⠏]")
_WORKING_RE = re.compile(r"thinking|working", re.IGNORECASE)


class QwenAdapter(Adapter):
    name = "qwen"
    instructions_filename = "QWEN.md"
    permission_bypass_args = ("-y",)
    submit_keys = ("Enter",)
    install_meta = InstallMeta(
        package_manager="npm",
        install_command=("npm", "install", "-g", "@qwen-code/qwen-code"),
        update_command=("npm", "install", "-g", "@qwen-code/qwen-code@latest"),
        version_command=("qwen", "--version"),
    )

    DEFAULT_MODEL = "qwen3-coder"

    def build_command(self, spec: RunSpec) -> BuildCommand:
        resolved = self.resolve_run_spec(spec)
        args = ["-p", spec.prompt, *resolved.permission_args, "-m", resolved.model, "--output-format", "json"]
        return self.finalize_command(spec, cmd="qwen", args=args)

    def parse_output(self, spec: RunSpec, outcome: SubprocOutcome) -> ParsedOutput:
        tokens_in, tokens_out, raw = _parse_qwen_stats(outcome.stdout)
        return {
            "cost_usd": None,
            "tokens_in": tokens_in if raw is not None else None,
            "tokens_out": tokens_out if raw is not None else None,
            "raw": raw,
        }

    # ---- session-aware ---------------------------------------------------

    def detect_ready(self, pane: str) -> ReadyState:
        last30 = last_non_empty_join(pane, 30)
        # Auth dialog (OAuth discontinued / API key prompt)
        if _AUTH_RE.search(last30) and _AUTH_ACTION_RE.search(last30):
            return "dialog"
        if _PROMPT_RE.search(last30):
            return "ready"
        return "loading"

    def handle_dialog(self, pane: str) -> list[str] | None:
        # Auth dialog needs the user — None so the consumer surfaces it.
        return None

    def detect_status(self, pane: str) -> AgentStatus:
        last10 = last_non_empty_join(pane, 10)
        if _RATE_LIMIT_RE.search(last10):
            return "rate-limited"
        if _SPINNER_RE.search(last10) or _WORKING_RE.search(last10):
            return "running"
        if _PROMPT_RE.search(last10):
            return "idle"
        return "unknown"

    def session_log_path(self, workdir: Path, session_started_after: float | None = None) -> str | None:
        roots = _project_roots(workdir)
        projects = _qwen_runtime_dir(workdir) / "projects"
        best: tuple[float, str, Path] | None = None
        for chats in dict.fromkeys(projects / _sanitize_cwd(root) / "chats" for root in roots):
            try:
                entries = list(chats.iterdir())
            except OSError:
                continue
            for entry in entries:
                if _SESSION_FILE_RE.match(entry.name) is None:
                    continue
                try:
                    if not entry.is_file():
                        continue
                    mtime = entry.stat().st_mtime
                except OSError:
                    continue
                if session_started_after is not None and mtime < session_started_after:
                    continue
                if best is not None and (mtime, entry.name) <= best[:2]:
                    continue
                if _session_cwd(entry) not in roots:
                    continue
                best = (mtime, entry.name, entry)
        return str(best[2]) if best is not None else None

    def parse_session_log(self, path: str) -> SessionTelemetry:
        p = Path(path)
        if not p.exists():
            return SessionTelemetry(path, None, None, None, None, None)
        try:
            raw_text = p.read_text(encoding="utf-8")
        except OSError:
            return SessionTelemetry(path, None, None, None, None, None)

        records = _qwen_records(raw_text)
        if records is not None:
            tokens_in, tokens_out, model = _records_usage(records)
            cost = derive_cost(model, tokens_in, tokens_out) if tokens_in is not None else None
            return SessionTelemetry(path, tokens_in, tokens_out, cost, model, records)

        stats = _parse_qwen_stats_blob(raw_text)
        if stats["tokens_in"] is not None and stats["tokens_out"] is not None:
            return SessionTelemetry(
                path,
                stats["tokens_in"],
                stats["tokens_out"],
                stats["cost_usd"],
                stats["model"],
                stats["raw"],
            )

        try:
            raw = json.loads(raw_text)
            return SessionTelemetry(path, None, None, None, None, raw)
        except json.JSONDecodeError:
            return SessionTelemetry(path, None, None, None, None, None)


# SessionService.SESSION_FILE_PATTERN: sidecars (.runtime.json, .ledger.jsonl, .pr.json) never match.
_SESSION_FILE_RE = re.compile(r"^[0-9a-fA-F-]{32,36}\.jsonl$")
_RUNTIME_SNAPSHOT_PREFIX = "$runtime|"
_MAX_SAFE_INTEGER = 9007199254740991


def _sanitize_cwd(cwd: str) -> str:
    # JavaScript's sanitizer replaces UTF-16 units, including both surrogate halves.
    return re.sub(r"[^a-zA-Z0-9]", lambda match: "--" if ord(match[0]) > 0xFFFF else "-", cwd)


def _resolve_qwen_path(value: str, cwd: Path) -> Path:
    """`Storage.resolvePath`: leading `~` expands to the home dir; relative paths resolve against the CLI cwd."""
    if value == "~" or value.startswith(("~/", "~\\")):
        return Path.home().joinpath(*[part for part in re.split(r"[/\\]+", value[2:]) if part])
    path = Path(value)
    return path if path.is_absolute() else Path(os.path.abspath(cwd / path))


def _qwen_runtime_dir(workdir: Path) -> Path:
    """`Storage.getRuntimeBaseDir()`: `QWEN_RUNTIME_DIR`, else `QWEN_HOME`, else `~/.qwen`.

    The settings-file `runtimeOutputDir` override is not consulted.
    """
    runtime = os.environ.get("QWEN_RUNTIME_DIR")
    if runtime:
        return _resolve_qwen_path(runtime, workdir)
    home = os.environ.get("QWEN_HOME")
    if home:
        return _resolve_qwen_path(home, workdir)
    return Path.home() / ".qwen"


def _project_roots(workdir: Path) -> list[str]:
    """Spellings qwen-code may have used as the project root (`path.resolve(process.cwd())`, physical path first)."""
    absolute = os.path.abspath(workdir)
    try:
        real = os.path.realpath(workdir)
    except OSError:
        real = absolute
    return [real] if real == absolute else [real, absolute]


def _session_cwd(path: Path) -> str | None:
    """`cwd` of the first record; qwen-code's own project-membership check reads the same field."""
    try:
        with path.open(encoding="utf-8") as handle:
            for line in handle:
                if line.strip():
                    record = json.loads(line)
                    break
            else:
                return None
    except (OSError, ValueError):
        return None
    cwd = record.get("cwd") if isinstance(record, dict) else None
    return cwd if isinstance(cwd, str) else None


def _qwen_records(text: str) -> list[dict] | None:
    """Parse a chats/ JSONL transcript; None when no line is a ChatRecord.

    Records are keyed by `uuid` (last write wins) so a re-emitted record is
    never counted twice.
    """
    records: dict[str, dict] = {}
    for index, line in enumerate(text.splitlines()):
        if not line.strip():
            continue
        try:
            record = json.loads(line)
        except ValueError:
            continue
        if not isinstance(record, dict) or not isinstance(record.get("type"), str) or not isinstance(record.get("sessionId"), str):
            continue
        uuid = record.get("uuid")
        records[uuid if isinstance(uuid, str) else f"line:{index}"] = record
    return list(records.values()) if records else None


def _usage_count(value: object) -> int | None:
    """Nonnegative safe integer; an absent field counts as 0 (upstream `?? 0`)."""
    if value is None:
        return 0
    if type(value) in (int, float) and 0 <= value <= _MAX_SAFE_INTEGER and int(value) == value:
        return int(value)
    return None


def _records_usage(records: list[dict]) -> tuple[int | None, int | None, str | None]:
    """Sum `usageMetadata.{promptTokenCount,candidatesTokenCount}` over assistant records.

    Only per-turn `usageMetadata` is counted; `ui_telemetry` and other system
    snapshots are cumulative views of the same usage and are ignored. None until
    any assistant record reports usage.
    """
    tokens_in: int | None = None
    tokens_out: int | None = None
    model: str | None = None
    usage_models: set[str | None] = set()
    for record in records:
        if record.get("type") != "assistant":
            continue
        record_model: str | None = None
        name = record.get("model")
        if isinstance(name, str) and name:
            while name.startswith(_RUNTIME_SNAPSHOT_PREFIX):
                stripped = "|".join(name.split("|")[2:])
                if not stripped:
                    break
                name = stripped
            model = name
            record_model = name
        usage = record.get("usageMetadata")
        if usage is None:
            continue
        if not isinstance(usage, dict):
            return None, None, model
        count_in = _usage_count(usage.get("promptTokenCount"))
        count_out = _usage_count(usage.get("candidatesTokenCount"))
        if count_in is None or count_out is None:
            return None, None, model
        usage_models.add(record_model)
        tokens_in = (tokens_in or 0) + count_in
        tokens_out = (tokens_out or 0) + count_out
        if tokens_in > _MAX_SAFE_INTEGER or tokens_out > _MAX_SAFE_INTEGER:
            return None, None, model
    if usage_models:
        model = next(iter(usage_models)) if len(usage_models) == 1 else None
    return tokens_in, tokens_out, model


def _parse_qwen_stats_blob(blob: str) -> dict:
    """Extract token/cost/model from a qwen stats envelope (`stats.models[*]`).

    Mirrors the TS `parseQwenStatsBlob`. Returns a dict with keys
    `tokens_in`, `tokens_out`, `cost_usd`, `model`, `raw`. Token fields are
    None when the blob can't be interpreted as a stats envelope.
    """
    try:
        parsed = json.loads(blob)
    except json.JSONDecodeError:
        return {"tokens_in": None, "tokens_out": None, "cost_usd": None, "model": None, "raw": None}
    return _stats_from_parsed(parsed)


def _stats_from_parsed(parsed: object) -> dict:
    if not isinstance(parsed, dict):
        return {"tokens_in": None, "tokens_out": None, "cost_usd": None, "model": None, "raw": parsed}

    stats = parsed.get("stats")
    models = stats.get("models") if isinstance(stats, dict) else None
    if not isinstance(models, dict):
        return {"tokens_in": None, "tokens_out": None, "cost_usd": None, "model": None, "raw": parsed}

    tokens_in = 0
    tokens_out = 0
    model: str | None = None
    for name, model_stats in models.items():
        if not isinstance(model_stats, dict):
            continue
        if model is None:
            model = name
        tokens = model_stats.get("tokens") or {}
        tokens_in += int(tokens.get("input") or 0)
        tokens_out += int(tokens.get("candidates") or 0)

    return {
        "tokens_in": tokens_in,
        "tokens_out": tokens_out,
        "cost_usd": derive_cost(model, tokens_in, tokens_out),
        "model": model,
        "raw": parsed,
    }


def _parse_qwen_stats(stdout: str) -> tuple[int, int, list | dict | None]:
    """Parse qwen's --output-format json output.

    Current qwen emits a JSON array where the last item with `type="result"`
    carries `usage.{input_tokens, output_tokens}`. Older qwen versions emit a
    JSON envelope object with `stats.models[*].tokens.{input, candidates}`;
    we keep a fallback for that for backwards compatibility.
    """
    candidates: list[str] = [stdout.strip()]
    for ln in stdout.splitlines():
        s = ln.strip()
        if s.startswith("[") or s.startswith("{"):
            candidates.append(s)

    for blob in candidates:
        if not blob:
            continue
        try:
            parsed = json.loads(blob)
        except json.JSONDecodeError:
            continue

        if isinstance(parsed, list):
            for item in reversed(parsed):
                if not isinstance(item, dict) or item.get("type") != "result":
                    continue
                usage = item.get("usage") or {}
                tokens_in = int(usage.get("input_tokens") or 0)
                tokens_out = int(usage.get("output_tokens") or 0)
                return tokens_in, tokens_out, parsed
            continue

        if isinstance(parsed, dict):
            stats = _stats_from_parsed(parsed)
            if stats["tokens_in"] is None:
                continue
            return stats["tokens_in"], stats["tokens_out"], parsed

    return 0, 0, None

File: ts/src/adapters/qwen.ts
import { register } from '../registry.js'
import type { Adapter, AgentStatus, BuildCommand, ParsedOutput, ReadyState, RunSpec, SessionTelemetry, SubprocOutcome } from '../base.js'
import { finalizeCommand, validateRunSpec } from '../base.js'
import { stripAnsi, lastNonEmptyJoin } from '../util.js'
import { existsSync, openSync, readSync, closeSync, readdirSync, readFileSync, realpathSync, statSync } from 'node:fs'
import { homedir } from 'node:os'
import { isAbsolute, join, resolve } from 'node:path'
import { deriveCost } from '../pricing.js'

type QwenStats = { tokensIn: number | null; tokensOut: number | null; costUsd: number | null; model: string | null; raw: unknown | null }

function parseQwenStatsBlob(blob: string): QwenStats {
  let parsed: unknown
  try {
    parsed = JSON.parse(blob)
  } catch {
    return { tokensIn: null, tokensOut: null, costUsd: null, model: null, raw: null }
  }
  return statsFromParsed(parsed)
}

function statsFromParsed(parsed: unknown): QwenStats {
  if (parsed === null || typeof parsed !== 'object' || Array.isArray(parsed)) {
    return { tokensIn: null, tokensOut: null, costUsd: null, model: null, raw: parsed }
  }
  const obj = parsed as Record<string, unknown>
  const stats = obj['stats'] as Record<string, unknown> | undefined
  const models = stats?.['models']
  if (!models || typeof models !== 'object') {
    return { tokensIn: null, tokensOut: null, costUsd: null, model: null, raw: parsed }
  }

  let tokensIn = 0
  let tokensOut = 0
  let model: string | null = null
  for (const [name, modelStats] of Object.entries(models as Record<string, unknown>)) {
    if (!modelStats || typeof modelStats !== 'object') continue
    if (!model) model = name
    const tokens = (modelStats as Record<string, unknown>)['tokens'] as Record<string, unknown> | undefined
    tokensIn += Number(tokens?.['input'] ?? 0)
    tokensOut += Number(tokens?.['candidates'] ?? 0)
  }

  return {
    tokensIn,
    tokensOut,
    costUsd: deriveCost(model, tokensIn, tokensOut),
    model,
    raw: parsed,
  }
}

// SessionService.SESSION_FILE_PATTERN: sidecars (.runtime.json, .ledger.jsonl, .pr.json) never match.
const SESSION_FILE_RE = /^[0-9a-fA-F-]{32,36}\.jsonl$/
const RUNTIME_SNAPSHOT_PREFIX = '$runtime|'

/** Parsed JSON as an object record, or null for scalars/arrays. */
function jsonObject(value: unknown): Record<string, unknown> | null {
  return value !== null && typeof value === 'object' && !Array.isArray(value) ? (value as Record<string, unknown>) : null
}

/** `Storage.resolvePath`: leading `~` expands to the home dir; relative paths resolve against the CLI cwd. */
function resolveQwenPath(value: string, cwd: string): string {
  if (value === '~' || value.startsWith('~/') || value.startsWith('~\\')) {
    return join(process.env.HOME ?? homedir(), ...value.slice(2).split(/[/\\]+/).filter(Boolean))
  }
  return isAbsolute(value) ? value : resolve(cwd, value)
}

/**
 * `Storage.getRuntimeBaseDir()`: `QWEN_RUNTIME_DIR`, else `QWEN_HOME`, else
 * `~/.qwen`. The settings-file `runtimeOutputDir` override is not consulted.
 */
function qwenRuntimeDir(workdir: string): string {
  const runtime = process.env['QWEN_RUNTIME_DIR']
  if (runtime) return resolveQwenPath(runtime, workdir)
  const home = process.env['QWEN_HOME']
  if (home) return resolveQwenPath(home, workdir)
  return join(process.env.HOME ?? homedir(), '.qwen')
}

/** Spellings qwen-code may have used as the project root (`path.resolve(process.cwd())`, physical path first). */
function projectRoots(workdir: string): string[] {
  const absolute = resolve(workdir)
  let real = absolute
  try {
    real = realpathSync(workdir)
  } catch {
    /* keep the literal path */
  }
  return real === absolute ? [real] : [real, absolute]
}

/** `cwd` of the first record; qwen-code's own project-membership check reads the same field. */
function sessionCwd(path: string): string | null {
  let first: string | undefined
  try {
    const fd = openSync(path, 'r')
    try {
      const buffer = Buffer.alloc(64 * 1024)
      const read = readSync(fd, buffer, 0, buffer.length, 0)
      first = buffer.toString('utf-8', 0, read).split('\n').find(line => line.trim())
    } finally {
      closeSync(fd)
    }
  } catch {
    return null
  }
  if (first === undefined) return null
  let record: unknown
  try {
    record = JSON.parse(first)
  } catch {
    return null
  }
  const cwd = jsonObject(record)?.['cwd']
  return typeof cwd === 'string' ? cwd : null
}

/**
 * Parse a chats/ JSONL transcript; null when no line is a ChatRecord. Records
 * are keyed by `uuid` (last write wins) so a re-emitted record is never counted twice.
 */
function qwenRecords(text: string): Record<string, unknown>[] | null {
  const records = new Map<string, Record<string, unknown>>()
  const lines = text.split('\n')
  for (let index = 0; index < lines.length; index++) {
    const line = lines[index]!
    if (!line.trim()) continue
    let parsed: unknown
    try {
      parsed = JSON.parse(line)
    } catch {
      continue
    }
    const record = jsonObject(parsed)
    if (!record || typeof record['type'] !== 'string' || typeof record['sessionId'] !== 'string') continue
    const uuid = record['uuid']
    records.set(typeof uuid === 'string' ? uuid : `line:${index}`, record)
  }
  return records.size > 0 ? [...records.values()] : null
}

/** Nonnegative safe integer; an absent field counts as 0 (upstream `?? 0`). */
function usageCount(value: unknown): number | null {
  if (value == null) return 0
  return typeof value === 'number' && Number.isSafeInteger(value) && value >= 0 ? value : null
}

/**
 * Sum `usageMetadata.{promptTokenCount,candidatesTokenCount}` over assistant
 * records. Only per-turn usageMetadata is counted; `ui_telemetry` and other
 * system snapshots are cumulative views of the same usage and are ignored.
 * Null until any assistant record reports usage.
 */
function recordsUsage(records: Record<string, unknown>[]): { tokensIn: number | null; tokensOut: number | null; model: string | null } {
  let tokensIn: number | null = null
  let tokensOut: number | null = null
  let model: string | null = null
  const usageModels = new Set<string | null>()
  for (const record of records) {
    if (record['type'] !== 'assistant') continue
    let recordModel: string | null = null
    if (typeof record['model'] === 'string' && record['model']) {
      let name: string = record['model']
      while (name.startsWith(RUNTIME_SNAPSHOT_PREFIX)) {
        const stripped = name.split('|').slice(2).join('|')
        if (!stripped) break
        name = stripped
      }
      model = name
      recordModel = name
    }
    if (record['usageMetadata'] == null) continue
    const usage = jsonObject(record['usageMetadata'])
    if (!usage) return { tokensIn: null, tokensOut: null, model }
    const countIn = usageCount(usage['promptTokenCount'])
    const countOut = usageCount(usage['candidatesTokenCount'])
    if (countIn === null || countOut === null) return { tokensIn: null, tokensOut: null, model }
    usageModels.add(recordModel)
    tokensIn = (tokensIn ?? 0) + countIn
    tokensOut = (tokensOut ?? 0) + countOut
    if (!Number.isSafeInteger(tokensIn) || !Number.isSafeInteger(tokensOut)) return { tokensIn: null, tokensOut: null, model }
  }
  if (usageModels.size) model = usageModels.size === 1 ? usageModels.values().next().value ?? null : null
  return { tokensIn, tokensOut, model }
}

const qwenAdapter: Adapter = {
  name: 'qwen',
  instructionsFilename: 'QWEN.md',
  defaultModel: 'qwen3-coder',
  permissionBypassArgs: ['-y'],

  buildCommand(spec: RunSpec): BuildCommand {
    const validated = validateRunSpec(this, spec)
    const { model, permissionArgs } = validated
    return finalizeCommand(this, spec, validated, {
      cmd: 'qwen',
      args: ['-p', spec.prompt, ...permissionArgs, '-m', model, '--output-format', 'json'],
    })
  },

  parseOutput(_spec: RunSpec, outcome: SubprocOutcome): ParsedOutput {
    // Current qwen emits a JSON array whose last `type: "result"` item carries
    // usage. Older versions emit a `stats.models[*]` envelope object; keep that
    // fallback so headless parsing matches the Python adapter.
    const candidates: string[] = [outcome.stdout.trim()]
    for (const ln of outcome.stdout.split('\n')) {
      const s = ln.trim()
      if (s.startsWith('[') || s.startsWith('{')) candidates.push(s)
    }

    for (const blob of candidates) {
      if (!blob) continue
      let parsed: unknown
      try {
        parsed = JSON.parse(blob)
      } catch {
        continue
      }
      if (!Array.isArray(parsed)) {
        const stats = statsFromParsed(parsed)
        if (stats.tokensIn === null) continue
        if (!Number.isFinite(stats.tokensIn) || !Number.isFinite(stats.tokensOut)) {
          throw new Error('Invalid Qwen token totals')
        }
        return { costUsd: null, tokensIn: stats.tokensIn, tokensOut: stats.tokensOut, raw: parsed }
      }
      for (let i = parsed.length - 1; i >= 0; i--) {
        const item = parsed[i]
        if (!item || typeof item !== 'object') continue
        const obj = item as Record<string, unknown>
        if (obj['type'] !== 'result') continue
        const usage = obj['usage'] as Record<string, unknown> | undefined
        const tokensIn = Number(usage?.['input_tokens'] ?? 0)
        const tokensOut = Number(usage?.['output_tokens'] ?? 0)
        return { costUsd: null, tokensIn, tokensOut, raw: parsed }
      }
    }

    return { costUsd: null, tokensIn: null, tokensOut: null, raw: null }
  },

  // qwen-code records interactive sessions as
  // <QWEN_RUNTIME_DIR|QWEN_HOME|~/.qwen>/projects/<sanitizeCwd(projectRoot)>/chats/<sessionId>.jsonl.
  // Sanitised paths can collide, so the first record's literal cwd decides membership.
  sessionLogPath(workdir: string, since?: number): string | null {
    const roots = projectRoots(workdir)
    const projects = join(qwenRuntimeDir(workdir), 'projects')
    let best: { mtime: number; name: string; path: string } | null = null
    for (const chats of new Set(roots.map(root => join(projects, root.replace(/[^a-zA-Z0-9]/g, '-'), 'chats')))) {
      let entries: string[]
      try {
        entries = readdirSync(chats)
      } catch {
        continue
      }
      for (const name of entries) {
        if (!SESSION_FILE_RE.test(name)) continue
        const path = join(chats, name)
        let mtime: number
        try {
          const stat = statSync(path)
          if (!stat.isFile()) continue
          mtime = stat.mtimeMs
        } catch {
          continue
        }
        if (since !== undefined && mtime < since) continue
        if (best && (mtime < best.mtime || (mtime === best.mtime && name <= best.name))) continue
        const cwd = sessionCwd(path)
        if (cwd === null || !roots.includes(cwd)) continue
        best = { mtime, name, path }
      }
    }
    return best?.path ?? null
  },

  parseSessionLog(path: string): SessionTelemetry {
    if (!existsSync(path)) {
      return { sessionLogPath: path, tokensIn: null, tokensOut: null, costUsd: null, model: null, raw: null }
    }
    let rawText: string
    try {
      rawText = readFileSync(path, 'utf-8')
    } catch {
      return { sessionLogPath: path, tokensIn: null, tokensOut: null, costUsd: null, model: null, raw: null }
    }
    const records = qwenRecords(rawText)
    if (records !== null) {
      const { tokensIn, tokensOut, model } = recordsUsage(records)
      const costUsd = tokensIn === null ? null : deriveCost(model, tokensIn, tokensOut)
      return { sessionLogPath: path, tokensIn, tokensOut, costUsd, model, raw: records }
    }
    const statsParsed = parseQwenStatsBlob(rawText)
    if (statsParsed.tokensIn != null && statsParsed.tokensOut != null) {
      return {
        sessionLogPath: path,
        tokensIn: statsParsed.tokensIn,
        tokensOut: statsParsed.tokensOut,
        costUsd: statsParsed.costUsd,
        model: statsParsed.model,
        raw: statsParsed.raw,
      }
    }
    try {
      return { sessionLogPath: path, tokensIn: null, tokensOut: null, costUsd: null, model: null, raw: JSON.parse(rawText) as unknown }
    } catch {
      return { sessionLogPath: path, tokensIn: null, tokensOut: null, costUsd: null, model: null, raw: null }
    }
  },
}

qwenAdapter.submitKeys = ['Enter']
qwenAdapter.detectReady = function (pane: string): ReadyState {
  const last30 = lastNonEmptyJoin(pane, 30)
  // Auth dialog (OAuth discontinued / API key prompt)
  if (/Qwen OAuth|API Key/i.test(last30) && /Discontinued|switch/i.test(last30)) return 'dialog'
  if (/Type your message|>\s*$|❯\s*$/m.test(last30)) return 'ready'
  return 'loading'
}
qwenAdapter.handleDialog = function (pane: string): string[] | null {
  // Auth dialog needs user — return null so flt surfaces it.
  return null
}
qwenAdapter.detectStatus = function (pane: string): AgentStatus {
  const last10 = lastNonEmptyJoin(pane, 10)
  if (/rate.?limit|quota/i.test(last10)) return 'rate-limited'
  if (/[⠋⠙⠹⠸⠼⠴⠦⠧⠇⠏]/.test(last10) || /thinking|working/i.test(last10)) return 'running'
  if (/Type your message|>\s*$|❯\s*$/m.test(last10)) return 'idle'
  return 'unknown'
}
qwenAdapter.installMeta = {
  packageManager: 'npm',
  installCommand: ['npm', 'install', '-g', '@qwen-code/qwen-code'],
  updateCommand: ['npm', 'install', '-g', '@qwen-code/qwen-code@latest'],
  versionCommand: ['qwen', '--version'],
}

register('qwen', qwenAdapter)
